kroxylicious-operator/src/main/java/io/kroxylicious/kubernetes/operator/ClusterService.java

```diff
@@ -49,82 +46,20 @@ static String serviceName(VirtualKafkaCluster cluster) {
-                                portNum -> portNum,
-                                portNum -> cluster.getMetadata().getName() + "-" + portNum,
-                                (v1, v2) -> {
-                                    throw new IllegalStateException();
-                                },
-                                TreeMap::new));
-            }
-        }
-        throw new IllegalArgumentException("Couldn't find cluster with name " + cluster.getMetadata().getName());
-    }
-
-    protected Service clusterService(KafkaProxy primary,
-                                     Context<KafkaProxy> context,
-                                     VirtualKafkaCluster cluster) {
-        // @formatter:off
-        var serviceSpecBuilder = new ServiceBuilder()
-                .withNewMetadata()
-                    .withName(serviceName(cluster))
-                    .withNamespace(primary.getMetadata().getNamespace())
-                    .addToLabels(standardLabels(primary))
-                    .addNewOwnerReferenceLike(ResourcesUtil.ownerReferenceTo(primary)).endOwnerReference()
-                .endMetadata()
-                .withNewSpec()
-                    .withSelector(ProxyDeployment.podLabels(primary));
-        for (var portNumEntry : clusterPorts( context, cluster).entrySet()) {
-            serviceSpecBuilder = serviceSpecBuilder
-                    .addNewPort()
-                        .withName(portNumEntry.getValue())
-                        .withPort(portNumEntry.getKey())
-                        .withTargetPort(new IntOrString(portNumEntry.getKey()))
-                        .withProtocol("TCP")
-                    .endPort();
-        }
-
-        return serviceSpecBuilder
-                .endSpec()
-                .build();
-        // @formatter:on
-    }
-
     @Override
     public Map<String, Service> desiredResources(
                                                  KafkaProxy primary,
                                                  Context<KafkaProxy> context) {
-        return ResourcesUtil.clustersInNameOrder(context)
+        List<VirtualKafkaCluster> clusters = ResourcesUtil.clustersInNameOrder(context).toList();
+        Set<KafkaProxyIngress> ingresses = context.getSecondaryResources(KafkaProxyIngress.class);
+        ProxyIngressLayout layout = ProxyIngressLayout.layout(primary, clusters, ingresses);
+        Map<String, Service> services = clusters.stream()
                 .filter(cluster -> !SharedKafkaProxyContext.isBroken(context, cluster))
+                .flatMap(cluster -> layout.clusterLayout(cluster).map(ProxyIngressLayout.VirtualClusterLayout::services).orElse(Stream.empty()))
                 .collect(Collectors.toMap(
-                        cluster -> cluster.getMetadata().getName(),
-                        cluster -> clusterService(primary, context, cluster)));
+                        service -> service.getMetadata().getName(),
```

Comment:
We do an awful lot of `thing.getMetadata().getName()`, that's also occurring an awful lot as a lambda expression. If we defined a `name(HasMetadata resource)` method we'd be able to use it as a method reference.

Reply:
done, there were also a couple of instances of this `indexByName` pattern that I pulled out a method for.